Failing E2E test  [sig-network] LoadBalancers ExternalTrafficPolicy: Local [Feature:LoadBalancer] [Slow] should only target nodes with endpoints

### Which jobs are failing?

 [sig-network] LoadBalancers ExternalTrafficPolicy: Local [Feature:LoadBalancer] [Slow] should only target nodes with endpoints

### Which tests are failing?

[PANICKED] in [It] - runtime/panic.go:115 @ 04/22/25 09:12:26.918
  I0422 09:12:26.918509 23 util.go:81]
  Output of kubectl describe svc:

  I0422 09:12:26.918605 23 builder.go:121] Running '/usr/local/bin/kubectl --kubeconfig=/tmp/kubeconfig-310162115 --namespace=esipp-9715 describ
e svc --namespace=esipp-9715'
  I0422 09:12:27.044900 23 builder.go:146] stderr: ""
  I0422 09:12:27.044964 23 builder.go:147] stdout: "Name:                     external-local-nodes\nNamespace:                esipp-9715\nLabels
:                   testid=external-local-nodes-73c595e0-4cce-4980-8804-44b8c33c8528\nAnnotations:              <none>\nSelector:
  testid=external-local-nodes-73c595e0-4cce-4980-8804-44b8c33c8528\nType:                     LoadBalancer\nIP Family Policy:         SingleStac
k\nIP Families:              IPv4\nIP:                       10.107.95.39\nIPs:                      10.107.95.39\nLoadBalancer Ingress:     10.
158.34.156 (VIP)\nPort:                     <unset>  8081/TCP\nTargetPort:               80/TCP\nNodePort:                 <unset>  30837/TCP\nE
ndpoints:                192.168.130.169:80\nSession Affinity:         None\nExternal Traffic Policy:  Local\nInternal Traffic Policy:  Cluster\
nHealthCheck NodePort:     30490\nEvents:\n  Type    Reason                Age   From                Message\n  ----    ------                --
--  ----                -------\n  Normal  EnsuringLoadBalancer  34s   service-controller  Ensuring load balancer\n  Normal  EnsuredLoadBalancer
   34s   service-controller  Ensured load balancer\n\n\nName:                     node-port-service\nNamespace:                esipp-9715\nLabel
s:                   <none>\nAnnotations:              <none>\nSelector:                 selector-abe458e2-85d6-44bb-a950-52754b3adf53=true\nTyp
e:                     NodePort\nIP Family Policy:         SingleStack\nIP Families:              IPv4\nIP:                       10.101.131.252
\nIPs:                      10.101.131.252\nPort:                     http  80/TCP\nTargetPort:               8083/TCP\nNodePort:
  http  32453/TCP\nEndpoints:                192.168.130.168:8083,192.168.196.139:8083,192.168.136.190:8083 + 1 more...\nPort:
   udp  90/UDP\nTargetPort:               8081/UDP\nNodePort:                 udp  31931/UDP\nEndpoints:                192.168.130.168:8081,192
.168.196.139:8081,192.168.136.190:8081 + 1 more...\nSession Affinity:         None\nExternal Traffic Policy:  Cluster\nInternal Traffic Policy:
 Cluster\nEvents:                   <none>\n\n\nName:                     session-affinity-service\nNamespace:                esipp-9715\nLabels
:                   <none>\nAnnotations:              <none>\nSelector:                 selector-abe458e2-85d6-44bb-a950-52754b3adf53=true\nType
:                     NodePort\nIP Family Policy:         SingleStack\nIP Families:              IPv4\nIP:                       10.97.15.200\nI
Ps:                      10.97.15.200\nPort:                     http  80/TCP\nTargetPort:               8083/TCP\nNodePort:                 htt
p  31513/TCP\nEndpoints:                192.168.130.168:8083,192.168.196.139:8083,192.168.136.190:8083 + 1 more...\nPort:                     ud
p  90/UDP\nTargetPort:               8081/UDP\nNodePort:                 udp  31478/UDP\nEndpoints:                192.168.130.168:8081,192.168.
196.139:8081,192.168.136.190:8081 + 1 more...\nSession Affinity:         ClientIP\nExternal Traffic Policy:  Cluster\nInternal Traffic Policy:
Cluster\nEvents:                   <none>\n"
  I0422 09:12:27.044976 23 util.go:84] Name:                     external-local-nodes
  Namespace:                esipp-9715
  Labels:                   testid=external-local-nodes-73c595e0-4cce-4980-8804-44b8c33c8528
  Annotations:              <none>
  Selector:                 testid=external-local-nodes-73c595e0-4cce-4980-8804-44b8c33c8528
  Type:                     LoadBalancer
  IP Family Policy:         SingleStack
  IP Families:              IPv4
  IP:                       10.107.95.39
  IPs:                      10.107.95.39
  LoadBalancer Ingress:     10.158.34.156 (VIP)
  Port:                     <unset>  8081/TCP
  TargetPort:               80/TCP
  NodePort:                 <unset>  30837/TCP
  Endpoints:                192.168.130.169:80
  Session Affinity:         None
  External Traffic Policy:  Local
  Internal Traffic Policy:  Cluster
  HealthCheck NodePort:     30490
  Events:
    Type    Reason                Age   From                Message
    ----    ------                ----  ----                -------
    Normal  EnsuringLoadBalancer  34s   service-controller  Ensuring load balancer
    Normal  EnsuredLoadBalancer   34s   service-controller  Ensured load balancer


  Name:                     node-port-service
  Namespace:                esipp-9715
  Labels:                   <none>
  Annotations:              <none>
  Selector:                 selector-abe458e2-85d6-44bb-a950-52754b3adf53=true
  Type:                     NodePort
  IP Family Policy:         SingleStack
  IP Families:              IPv4
  IP:                       10.101.131.252
  IPs:                      10.101.131.252
  Port:                     http  80/TCP
  TargetPort:               8083/TCP
  NodePort:                 http  32453/TCP
  Endpoints:                192.168.130.168:8083,192.168.196.139:8083,192.168.136.190:8083 + 1 more...
  Port:                     udp  90/UDP
  TargetPort:               8081/UDP
  NodePort:                 udp  31931/UDP
  Endpoints:                192.168.130.168:8081,192.168.196.139:8081,192.168.136.190:8081 + 1 more...
  Session Affinity:         None
  External Traffic Policy:  Cluster
  Internal Traffic Policy:  Cluster
  Events:                   <none>


  Name:                     session-affinity-service
  Namespace:                esipp-9715
  Labels:                   <none>
  Annotations:              <none>
  Selector:                 selector-abe458e2-85d6-44bb-a950-52754b3adf53=true
  Type:                     NodePort
  IP Family Policy:         SingleStack
  IP Families:              IPv4
  IP:                       10.97.15.200
  IPs:                      10.97.15.200
  Port:                     http  80/TCP
  TargetPort:               8083/TCP
  NodePort:                 http  31513/TCP
  Endpoints:                192.168.130.168:8083,192.168.196.139:8083,192.168.136.190:8083 + 1 more...
  Port:                     udp  90/UDP
  TargetPort:               8081/UDP
  NodePort:                 udp  31478/UDP
  Endpoints:                192.168.130.168:8081,192.168.196.139:8081,192.168.136.190:8081 + 1 more...
  Session Affinity:         ClientIP
  External Traffic Policy:  Cluster
  Internal Traffic Policy:  Cluster
  Events:                   <none>

  I0422 09:12:27.066184 23 jig.go:604] Waiting up to 15m0s for service "external-local-nodes" to have no LoadBalancer
  I0422 09:12:27.092657 23 helper.go:125] Waiting up to 7m0s for all (but 0) nodes to be ready
  STEP: dump namespace information after failure @ 04/22/25 09:12:27.097

................................................................................................................................................................


  [PANICKED] Test Panicked
  In [It] at: runtime/panic.go:115 @ 04/22/25 09:12:26.918

  runtime error: index out of range [3] with length 3

  Full Stack Trace
    k8s.io/kubernetes/test/e2e/network.init.func16.4({0x7f9d2c2f99e0, 0xc00527e510})
        k8s.io/kubernetes/test/e2e/network/loadbalancer.go:1136 +0xb25
------------------------------
......................................................
Summarizing 1 Failure:
  [PANICKED!] [sig-network] LoadBalancers ExternalTrafficPolicy: Local [Feature:LoadBalancer] [Slow] [It] should only target nodes with endpoint
s [sig-network, Feature:LoadBalancer, Slow]
  runtime/panic.go:115

Ran 1 of 6622 Specs in 34.664 seconds
FAIL! -- 0 Passed | 1 Failed | 0 Pending | 6621 Skipped
--- FAIL: TestE2E (34.95s)
FAIL


### Since when has it been failing?

The test failed 04/2025

### Testgrid link

_No response_

### Reason for failure (if possible)

_No response_

### Anything else we need to know?

_No response_

### Relevant SIG(s)

/sig-network

--- a/test/e2e/network/loadbalancer.go
+++ b/test/e2e/network/loadbalancer.go
@@ -1103,7 +1103,20 @@ var _ = common.SIGDescribe("LoadBalancers ExternalTrafficPolicy: Local", feature
 			framework.Failf("Service HealthCheck NodePort was not allocated")
 		}
 
-		ips := e2enode.CollectAddresses(nodes, v1.NodeInternalIP)
+		// Collect InternalIP addresses for each node.
+		nodeInternalIPs := make(map[string]sets.Set[string], len(nodes.Items))
+		for i := range nodes.Items {
+			node := &nodes.Items[i]
+			internalIPs := sets.New[string]()
+			for _, addr := range node.Status.Addresses {
+				if addr.Type == v1.NodeInternalIP && addr.Address != "" {
+					internalIPs.Insert(addr.Address)
+				}
+			}
+			if internalIPs.Len() > 0 {
+				nodeInternalIPs[node.Name] = internalIPs
+			}
+		}
 
 		ingressIP := e2eservice.GetIngressPoint(&svc.Status.LoadBalancer.Ingress[0])
 		svcTCPPort := int(svc.Spec.Ports[0].Port)
@@ -1128,22 +1141,24 @@ var _ = common.SIGDescribe("LoadBalancers ExternalTrafficPolicy: Local", feature
 
 			// HealthCheck should pass only on the node where num(endpoints) > 0
 			// All other nodes should fail the healthcheck on the service healthCheckNodePort
-			for n, internalIP := range ips {
-				// Make sure the loadbalancer picked up the health check change.
-				// Confirm traffic can reach backend through LB before checking healthcheck nodeport.
-				e2eservice.TestReachableHTTP(ctx, ingressIP, svcTCPPort, e2eservice.KubeProxyLagTimeout)
-				expectedSuccess := nodes.Items[n].Name == endpointNodeName
-				port := strconv.Itoa(healthCheckNodePort)
-				ipPort := net.JoinHostPort(internalIP, port)
-				framework.Logf("Health checking %s, http://%s/healthz, expectedSuccess %v", nodes.Items[n].Name, ipPort, expectedSuccess)
-				err := testHTTPHealthCheckNodePortFromTestContainer(ctx,
-					config,
-					internalIP,
-					healthCheckNodePort,
-					e2eservice.KubeProxyEndpointLagTimeout,
-					expectedSuccess,
-					threshold)
-				framework.ExpectNoError(err)
+			for n, internalIPs := range nodeInternalIPs {
+				for ip := range internalIPs {
+					// Make sure the loadbalancer picked up the health check change.
+					// Confirm traffic can reach backend through LB before checking healthcheck nodeport.
+					e2eservice.TestReachableHTTP(ctx, ingressIP, svcTCPPort, e2eservice.KubeProxyLagTimeout)
+					expectedSuccess := n == endpointNodeName
+					port := strconv.Itoa(healthCheckNodePort)
+					ipPort := net.JoinHostPort(ip, port)
+					framework.Logf("Health checking %s, http://%s/healthz, expectedSuccess %v", n, ipPort, expectedSuccess)
+					err := testHTTPHealthCheckNodePortFromTestContainer(ctx,
+						config,
+						ip,
+						healthCheckNodePort,
+						e2eservice.KubeProxyEndpointLagTimeout,
+						expectedSuccess,
+						threshold)
+					framework.ExpectNoError(err)
+				}
 			}
 
 			err = f.ClientSet.AppsV1().Deployments(namespace).Delete(ctx, serviceName, metav1.DeleteOptions{})

